qiskit/algorithms/quantum_time_evolution/real/implementations/trotterization/trotter_ops_validator.py

```diff
@@ -0,0 +1,156 @@
+
+import numbers
+from typing import Union
+
+from qiskit.circuit import Parameter, ParameterExpression
+from qiskit.opflow import (
+    OperatorBase,
+    StateFn,
+    SummedOp,
+    PauliOp,
+)
+
+
+def _is_op_bound(operator: Union[SummedOp, PauliOp, OperatorBase]) -> None:
+    """Checks if an operator provided has all parameters bound.
+
+    Args:
+        operator: Operator to be checked.
+
+    Raises:
+        ValueError: If an operator has unbound parameters.
+    """
+    if len(operator.parameters) > 0:
+        raise ValueError(
+            f"Did not manage to bind all parameters in the Hamiltonian, "
+            f"these parameters encountered: {operator.parameters}."
+        )
+
+
+def _validate_input(initial_state: StateFn, observable: OperatorBase) -> None:
+    """
+    Validates if one and only one among initial_state and observable is provided.
+
+    Args:
+        initial_state: A variable potentially holding a quantum state.
+        observable: A variable potentially holding a quantum observable.
+
+    Raises:
+        ValueError: If both initial_state and observable are provided or none of them.
+    """
+    if initial_state is None and observable is None:
+        raise ValueError(
+            "TrotterQrte requires an initial state or an observable to be evolved; None "
+            "provided."
+        )
+    if initial_state is not None and observable is not None:
+        raise ValueError(
+            "TrotterQrte requires an initial state or an observable to be evolved; both "
+            "provided."
+        )
+
+
+def _validate_hamiltonian_form(hamiltonian: Union[SummedOp, PauliOp, OperatorBase]):
+    """Validates that a Hamiltonian is of a correct type and with expected dependence on
+    parameters.
+
+    Args:
+        hamiltonian: Hamiltonian to be validated.
+
+    Raises:
+        ValueError: if an invalid Hamiltonian is provided.
+    """
+    if isinstance(hamiltonian, SummedOp):
+        if isinstance(hamiltonian.coeff, ParameterExpression):
+            raise ValueError(
+                "The coefficient multiplying the whole Hamiltonian cannot be a "
+                "ParameterExpression."
+            )
+        for op in hamiltonian.oplist:
+            if not _is_pauli_lin_single_param(op):
+                raise ValueError(
+                    "Hamiltonian term has a coefficient that is not a linear function of a "
+                    "single parameter. It is not supported."
+                )
+    elif isinstance(hamiltonian, (PauliOp, OperatorBase)):
+        if not _is_pauli_lin_single_param(hamiltonian):
+            raise ValueError(
+                "Hamiltonian term has a coefficient that is not a linear function of a "
+                "single parameter. It is not supported."
+            )
+    else:
+        raise ValueError("Hamiltonian not a SummedOp/PauliOp which are the only options supported.")
+
+
+def _is_pauli_lin_single_param(operator: PauliOp) -> bool:
```

Comment:
I understand that this is what you mention in the TO-DO list, where some checks are still not extended to hamiltonians of `OperatorBase` class, right? (Else, this probably should be pointed out)